### PR TITLE
Add a base for implementing caliper and xAPI processor and backends

### PR DESCRIPTION
**Pull request by danialmalik**
### PR description
We can use these changes and class by applying the following configurations in edx-platform.
**UPDATE**
Updating the settings below to match the current code in PR
```

REALTIME_EVENTING_BACKEND = {      # our
    'ENGINE': 'eventtracking.backends.routing.RoutingBackend',
    'OPTIONS': {
        'processors': [
            {
                'ENGINE': 'eventtracking.processors.filters.RegExFilterProcessor',
                'OPTIONS': {}
            }
        ],
        'backends': {
            'async_realtime_eventing': {
                'ENGINE': 'eventtracking.backends.routing.AsyncRoutingBackend',
                'OPTIONS': {
                    'backends': {
                        'caliper': {
                            'ENGINE': 'eventtracking.backends.caliper.CaliperBackend',
                            'OPTIONS': {
                                'processors': [
                                    #                 {'ENGINE':'translator'},
                                    #                 {'ENGINE':'validator'},
                                    #                 {'ENGINE':'filter'},
                                    #                 {'ENGINE':'router'}
                                ]
                            }
                        },
                        'xAPI': {
                            'ENGINE': 'eventtracking.backends.xapi.XApiBackend',
                            'OPTIONS': {
                                'processors': [
                                    #                 {'ENGINE':'translator'},
                                    #                 {'ENGINE':'validator'},
                                    #                 {'ENGINE':'filter'},
                                    #                 {'ENGINE':'router'}
                                ]
                            }
                        }
                    },
                    'processors': []
                }
            }
        }
    }
}

```
The settings above can be used in the `EVENT_TRACKING_BACKENDS` in the following way
```

EVENT_TRACKING_BACKENDS = {
    'tracking_logs': {
        'ENGINE': 'eventtracking.backends.routing.RoutingBackend',
        'OPTIONS': {
            'backends': {
                'logger': {
                    'ENGINE': 'eventtracking.backends.logger.LoggerBackend',
                    'OPTIONS': {
                        'name': 'tracking',
                        'max_event_size': TRACK_MAX_EVENT,
                    }
                },
                'realtime_eventtracking': REALTIME_EVENTING_BACKEND # OUR BACKEND SETTINGS
            },
            'processors': [
                {'ENGINE': 'track.shim.LegacyFieldMappingProcessor'},
                {'ENGINE': 'track.shim.PrefixedEventProcessor'}
            ]
        }
    },
   # other backends
}
```

### CODE REVIEW

**Comment by ziafazal** [comment]
@danialmalik this is how our setting should look like when used in django application(edx-platform)
```
EVENT_TRACKING_BACKENDS = {
    'tracking_logs': {
        'ENGINE': 'eventtracking.backends.routing.RoutingBackend',
        'OPTIONS': {
            .............
            },
            'processors': [
                .............
            ]
        }
    },
    'segmentio': {
        'ENGINE': 'eventtracking.backends.routing.RoutingBackend',
        'OPTIONS': {
            .............
            },
            'processors': [
                .............
            ]
        }
    },
    'caliper_events': {
        'ENGINE': 'eventtracking.backends.routing.AsyncRoutingBackend',
        'OPTIONS': {
            'backends': {
                'caliper': {
                    'ENGINE': 'eventtracking.backends.caliper.CaliperBackend',
                    'OPTIONS': {
                        'Caliper backend related options if any'
                    }
                }
            },
            'processors': [
                {'ENGINE': 'eventtracking.processor.caliper.CaliperEventProcessor'},
            ]
        }
    },
    'xapi_events': {
        'ENGINE': 'eventtracking.backends.routing.AsyncRoutingBackend',
        'OPTIONS': {
            'backends': {
                'xapi': {
                    'ENGINE': 'eventtracking.backends.xapi.XapiBackend',
                    'OPTIONS': {
                        'Xapi backend related options if any'
                    }
                }
            },
            'processors': [
                {'ENGINE': 'eventtracking.processor.xapi.XapiEventProcessor'},
            ]
        }
    },
}

```

File: eventtracking/backends/async_routing.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from pickle import PicklingError
+
+from eventtracking.processors.exceptions import EventEmissionExit
+from eventtracking.backends.routing import RoutingBackend
+from eventtracking.tasks import send_task_to_backend
+
+LOG = logging.getLogger(__name__)
+
+
+class AsyncRoutingBackend(RoutingBackend):
```
Comment (ziafazal) [inline eventtracking/backends/async_routing.py]: I think this should be part of existing `eventtracking/backends/routing.py`

File: eventtracking/backends/async_routing.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    def send(self, event):
```
Comment (ziafazal) [inline eventtracking/backends/async_routing.py]: this method should send event asynchronously
```
 try:
                pickled_routing_backend = pickle.dumps(self)
                ansyc_send.delay(pickled_routing_backend, event)
except PicklingError:
                LOG.exception(
                    'Unable to pickle AsyncRoutingBackend'
                )
```

File: eventtracking/backends/async_routing.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        Logs and swallows all `Exception`.
+        """
+        try:
+            processed_event = self.process_event(event)
+        except EventEmissionExit:
+            return
+
+        self.send_to_backends_async(processed_event)
+
+    def send_to_backends_async(self, event):
```
Comment (ziafazal) [inline eventtracking/backends/async_routing.py]: We dont need to override `send_to_backends` method.

File: eventtracking/caliper/backends.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import logging
+from celery.utils.log import get_task_logger
+
+from eventtracking.backends.routing import RoutingBackend
+
+LOGGER = get_task_logger(__name__)
+
+
+class CaliperBackend(RoutingBackend):
```
Comment (ziafazal) [inline eventtracking/caliper/backends.py]: why we need to inherit from `RoutingBackend`?

File: eventtracking/tasks.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+from importlib import import_module
+from celery.utils.log import get_task_logger
+
+from celery import task
+
+LOG = get_task_logger(__name__)
+
+
+@task(name='eventtracking.async_routing')
+def send_task_to_backend(backend_name, pickled_backend, event):
```
Comment (ziafazal) [inline eventtracking/tasks.py]: need to update it's signature
`ansyc_send(pickled_backend, event):`

File: eventtracking/xAPI/backends.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import logging
```
Comment (ziafazal) [inline eventtracking/xAPI/backends.py]: This file should be moved into `eventtracking/backends` and rename it xapi.

File: eventtracking/caliper/backends.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import logging
```
Comment (ziafazal) [inline eventtracking/caliper/backends.py]: This file should be moved into `eventtracking/backends` and rename it caliper.

File: eventtracking/xAPI/backends.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+from celery.utils.log import get_task_logger
+
+from eventtracking.backends.routing import RoutingBackend
+
+LOGGER = get_task_logger(__name__)
+
+
+class XApiBackend(RoutingBackend):
```
Comment (ziafazal) [inline eventtracking/xAPI/backends.py]: why we need to inherit from `RoutingBackend`?

**Review by ziafazal — CHANGES_REQUESTED** [review]
@danialmalik completed first round of review.

**Comment by danialmalik** [comment]
> @danialmalik this is how our setting should look like when used in django application(edx-platform)
> 
> ```
> EVENT_TRACKING_BACKENDS = {
>     'tracking_logs': {
>         'ENGINE': 'eventtracking.backends.routing.RoutingBackend',
>         'OPTIONS': {
>             .............
>             },
>             'processors': [
>                 .............
>             ]
>         }
>     },
>     'segmentio': {
>         'ENGINE': 'eventtracking.backends.routing.RoutingBackend',
>         'OPTIONS': {
>             .............
>             },
>             'processors': [
>                 .............
>             ]
>         }
>     },
>     'caliper_events': {
>         'ENGINE': 'eventtracking.backends.routing.AsyncRoutingBackend',
>         'OPTIONS': {
>             'backends': {
>                 'caliper': {
>                     'ENGINE': 'eventtracking.backends.caliper.CaliperBackend',
>                     'OPTIONS': {
>                         'Caliper backend related options if any'
>                     }
>                 }
>             },
>             'processors': [
>                 {'ENGINE': 'eventtracking.processor.caliper.CaliperEventProcessor'},
>             ]
>         }
>     },
>     'xapi_events': {
>         'ENGINE': 'eventtracking.backends.routing.AsyncRoutingBackend',
>         'OPTIONS': {
>             'backends': {
>                 'xapi': {
>                     'ENGINE': 'eventtracking.backends.xapi.XapiBackend',
>                     'OPTIONS': {
>                         'Xapi backend related options if any'
>                     }
>                 }
>             },
>             'processors': [
>                 {'ENGINE': 'eventtracking.processor.xapi.XapiEventProcessor'},
>             ]
>         }
>     },
> }
> ```
@ziafazal I've tried both the settings (one where caliper backend is inside the `tracking_logs` tree, and one where it's added as a sibling to `tracking_logs`). As we discussed, there is a vast difference in the JSON structure of the events that get passed to the caliper backend.
Attaching the SS
image

File: eventtracking/xAPI/backends.py (hunk 8 above)
Comment (danialmalik) [reply eventtracking/xAPI/backends.py]: `RoutingBackend` validates the processors and backends on initialization. Also, It has the method `process_event` that passes the events through the configured processors. 
Therefore inheriting from this Backend might save us from some effort.

File: eventtracking/caliper/backends.py (hunk 4 above)
Comment (danialmalik) [reply eventtracking/caliper/backends.py]: `RoutingBackend` validates the processors and backends on initialization. Also, It has the method `process_event` that passes the events through the configured processors. 
Therefore inheriting from this Backend might save us from some effort.

**Comment by danialmalik** [comment]
@ziafazal I've implemented the feedback. Please have another look at your time of leisure.

**Comment by ziafazal** [comment]
@danialmalik the difference in event is due to the context processor. if we add this to caliper backend then event should be same
`            'processors': [
                {'ENGINE': 'track.shim.LegacyFieldMappingProcessor'},
            ]
`
However we should not do that we should consume the events in the form they are emitted.

File: eventtracking/backends/routing.py [hunk 9]
```diff
@@ -134,3 +138,30 @@ def send_to_backends(self, event):
...
+        except EventEmissionExit:
+            return
+
+        for name, backend in six.iteritems(self.backends):
+            try:
+                pickled_backend = pickle.dumps(backend)
+                async_send.delay(name, pickled_backend, event)
+            except PicklingError:
+                LOG.exception(
+                    'Unable to pickle backend: %s', name
```
Comment (ziafazal) [inline eventtracking/backends/routing.py:162]: name should be replace with `backend`

File: eventtracking/backends/routing.py [hunk 10]
```diff
@@ -134,3 +138,30 @@ def send_to_backends(self, event):
...
+                )
+            except Exception:  # pylint: disable=broad-except
+                LOG.exception(
+                    'Unable to schedule task to send event to backend %s', name
```
Comment (ziafazal) [inline eventtracking/backends/routing.py:166]: name should be replace with `backend`

File: eventtracking/backends/xapi.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import logging
+from celery.utils.log import get_task_logger
+
+from eventtracking.backends.routing import RoutingBackend
+
+LOGGER = get_task_logger(__name__)
+
+
+class XApiBackend(RoutingBackend):
```
Comment (ziafazal) [inline eventtracking/backends/xapi.py:9]: Have you been able to figure out why we need to inherit `xapi` and `caliper` backends from `RoutingBackend`?

**Review by ziafazal — CHANGES_REQUESTED** [review]
@danialmalik completed another round of review. It looks better now. Few small things and a question.

File: eventtracking/backends/xapi.py (hunk 11 above)
Comment (danialmalik) [reply eventtracking/backends/xapi.py:9]: I've replied about that to the comments above.

File: eventtracking/backends/routing.py (hunk 9 above)
Comment (danialmalik) [reply eventtracking/backends/routing.py:162]: Using `backend` will result in logs like this: `Unable to schedule task to send event to backend <eventtracking.backends.caliper.CaliperBackend object at 0x7f154cb5cac8>`
Is this acceptable?

Comment (ziafazal) [reply eventtracking/backends/routing.py:162]: @danialmalik that's ugly then keep using `name`

File: eventtracking/caliper/backends.py (hunk 4 above)
Comment (ziafazal) [reply eventtracking/caliper/backends.py]: @danialmalik event is going to be processed by `AsyncRoutingBackend` before it is going to reach `CaliperBackend` or `xAPIBackend`. So it does not make sense to inherit them from `RoutingBackend`.

**Review by ziafazal — APPROVED** [review]

**Pull request merged**